gce/survival-training/README.md

```diff
@@ -0,0 +1,93 @@
+
+There are five components to our resilient training jobs:
+
+1. [GCE custom images](https://cloud.google.com/compute/docs/images#custom_images)
+
+1. A trainer command line interface
+
+1. [GCE startup scripts](https://cloud.google.com/compute/docs/startupscript)
+
+1. [GCE instance metadata](https://cloud.google.com/compute/docs/storing-retrieving-metadata)
+
+1. [Cloud Source Repositories](https://cloud.google.com/source-repositories/)
+
+
+As a part of this process, if you have not already done so, you will be prompted to enable both the Compute Engine API as well as the Cloud Source Repositories API for your Google Cloud Platform project.
+
+We will be working through the [Google Cloud SDK](https://cloud.google.com/sdk/). If you have not already done so, it is worth installing.
+
+
+### Custom images
+
+Your training job may require specific versions of libraries, like [tensorflow](https://www.tensorflow.org/) or [scikit-learn](http://scikit-learn.org). Additionally, in order to take advantage of instances with attached GPUs, you will have to have hardware-specific drivers installed on the instance. [Custom images](https://cloud.google.com/compute/docs/images#custom_images) are a convenient solution to this problem of environment reproduction within a GCE virtual machine.
+
+If you do have requirements (in addition to simply the OS) in your training environments, it is highly recommended that you create a VM image that you can use to spawn new instances without having to go through the tedious configuration process every time. The examples following this strategy section will demonstrate the creation of such images so, if this seems daunting, don't worry.
+
+
+### Trainer CLI
+
+Our strategy will have the GCE instance automatically kick off your training job on startup. As part of this kick off, it will have to provide the job with the appropriate parameters -- how many steps to train for, how often to checkpoint the model, which data to train the model with, what hyperparameters should be used in training the model, etc.
+
+It is helpful, if you intend to do this more than once, to assert a common semantical system across all your trainers. We provide a [boilerplate CLI](./dummy/train.py) in this repo. This interface is very similar to that provided by [TensorFlow estimators](https://www.tensorflow.org/programmers_guide/estimators), but can be used even with other frameworks. In the sections that follow, we will provide you with both TensorFlow and with scikit-learn examples that demonstrate its use.
+
+
+### Cloud Source Repositories
+
+We need a mechanism by which we can get our trainer code into the GCE instance. There are several solutions available to us:
+
+1. We could simply store the trainer application on the custom image from which we create the GCE instance.
+
+1. We could store the code in a [GCS](https://cloud.google.com/storage/) bucket and have the startup script download it to the instance before doing anything else.
```

Comment:
Actually a better pattern is to be able to consume a deployed package from GCS because it lets you decouple the build, deploy and versioning of the model code. This is nice because you can now assume the user will somehow (however (s)he wants) get the code to GCS and everything will work. Also you don't have to worry about how the user gets the model code to GCS which can be part of any build and deploy system.